ctrd/client.go

```diff
@@ -166,3 +187,132 @@ func (c *Client) Version(ctx context.Context) (containerd.Version, error) {
 
 	return cli.client.Version(ctx)
 }
+
+func (c *Client) runContainerdDaemon(homeDir string, copts clientOpts) error {
+	if homeDir == "" {
+		return fmt.Errorf("ctrd: containerd home dir should not be empty")
+	}
+
+	containerdPath, err := exec.LookPath(copts.containerdBinary)
+	if err != nil {
+		return fmt.Errorf("failed to find containerd binary %s: %v", copts.containerdBinary, err)
+	}
+
+	stateDir := path.Join(homeDir, "containerd/state")
+	if _, err := os.Stat(stateDir); err != nil && os.IsNotExist(err) {
+		if err := os.MkdirAll(stateDir, 0666); err != nil {
+			return fmt.Errorf("failed to mkdir %s: %v", stateDir, err)
+		}
+	}
+
+	pidFileName := path.Join(stateDir, containerdPidFileName)
+	f, err := os.OpenFile(pidFileName, os.O_RDWR|os.O_CREATE, 0600)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+
+	buf := make([]byte, 8)
+	num, err := f.Read(buf)
+	if err != nil && err != io.EOF {
+		return err
+	}
+
+	if num > 0 {
+		pid, err := strconv.ParseUint(string(buf[:num]), 10, 64)
+		if err != nil {
+			return err
+		}
+		if utils.IsProcessAlive(int(pid)) {
+			logrus.Infof("ctrd: previous instance of containerd still alive (%d)", pid)
+			c.daemonPid = int(pid)
+			return nil
+		}
+	}
+
+	// empty container pid file
+	_, err = f.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	if err := f.Truncate(0); err != nil {
+		return err
+	}
+
+	// if socket file exists, delete it.
+	if _, err := os.Stat(c.socketPath); err == nil {
+		os.RemoveAll(c.socketPath)
+	}
+
+	// Start a new containerd instance
```

Comment:
I think we can use a function to wrap the following functions

Reply:
+1